contracts/contracts/strategies/BaseCurveAMOStrategy.sol

```diff
@@ -0,0 +1,662 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+/**
+ * @title Curve Automated Market Maker (AMO) Strategy
+ * @notice AMO strategy for the Curve OETH/WETH pool
+ * @author Origin Protocol Inc
+ */
+import { Math } from "@openzeppelin/contracts/utils/math/Math.sol";
+import { SafeCast } from "@openzeppelin/contracts/utils/math/SafeCast.sol";
+
+import { IERC20, InitializableAbstractStrategy } from "../utils/InitializableAbstractStrategy.sol";
+import { StableMath } from "../utils/StableMath.sol";
+import { IVault } from "../interfaces/IVault.sol";
+import { IWETH9 } from "../interfaces/IWETH9.sol";
+import { ICurveStableSwapNG } from "../interfaces/ICurveStableSwapNG.sol";
+import { ICurveXChainLiquidityGauge } from "../interfaces/ICurveXChainLiquidityGauge.sol";
+import { IChildLiquidityGaugeFactory } from "../interfaces/IChildLiquidityGaugeFactory.sol";
+
+contract BaseCurveAMOStrategy is InitializableAbstractStrategy {
+    using StableMath for uint256;
+    using SafeCast for uint256;
+
+    /**
+     * @dev a threshold under which the contract no longer allows for the protocol to manually rebalance.
+     *      Guarding against a strategist / guardian being taken over and with multiple transactions
+     *      draining the protocol funds.
+     */
+    uint256 public constant SOLVENCY_THRESHOLD = 0.998 ether;
+
+    // New immutable variables that must be set in the constructor
+    /**
+     * @notice Address of the Wrapped ETH (WETH) contract.
+     */
+    IWETH9 public immutable weth;
+
+    /**
+     * @notice Address of the OETH token contract.
+     */
+    IERC20 public immutable oeth;
+
+    /**
+     * @notice Address of the LP (Liquidity Provider) token contract.
+     */
+    IERC20 public immutable lpToken;
+
+    /**
+     * @notice Address of the Curve StableSwap NG pool contract.
+     */
+    ICurveStableSwapNG public immutable curvePool;
+
+    /**
+     * @notice Address of the Curve X-Chain Liquidity Gauge contract.
+     */
+    ICurveXChainLiquidityGauge public immutable gauge;
+
+    /**
+     * @notice Address of the Child Liquidity Gauge Factory contract.
+     */
+    IChildLiquidityGaugeFactory public immutable gaugeFactory;
+
+    // Ordered list of pool assets
+    uint128 public constant oethCoinIndex = 1;
+    uint128 public constant ethCoinIndex = 0;
```

Comment:
confirming the order is correct:
<img width="457" alt="Screenshot 2025-02-07 at 09 56 16" src="https://github.com/user-attachments/assets/75aa53bd-6d15-48b5-afd8-d5222009ae6a" />
<img width="419" alt="Screenshot 2025-02-07 at 09 56 05" src="https://github.com/user-attachments/assets/3832162f-711c-49a0-8d57-45248ce90618" />